grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java

```diff
@@ -312,21 +372,29 @@ public T subType(T type){
         return getThis();
     }
 
-    private void checkForLoop(Schema.EdgeLabel edge){
+    private boolean superTypeLoops(){
         //Check For Loop
         HashSet<Type> foundTypes = new HashSet<>();
         Type currentSuperType = superType();
         while (currentSuperType != null){
             foundTypes.add(currentSuperType);
             currentSuperType = currentSuperType.superType();
             if(foundTypes.contains(currentSuperType)){
-                throw new ConceptException(ErrorMessage.LOOP_DETECTED.getMessage(toString(), edge.getLabel()));
+                return true;
             }
         }
+        return false;
     }
 
     T playsRole(RoleType roleType, boolean required) {
         checkTypeMutation();
+
+        //Update the internal cache of role types played
+        cachedPlaysRoles.get().add(roleType);
```

Comment:
I think you are correct here. In general the cache should only be updated if it has been loaded.